source/extensions/clusters/dns/dns_cluster.h

```diff
@@ -31,6 +31,73 @@ class DnsClusterFactory : public Upstream::ConfigurableClusterFactoryBase<
                           Upstream::ClusterFactoryContext& context) override;
 };
 
+class DnsClusterImpl : public BaseDynamicClusterImpl {
+public:
+  // Upstream::Cluster
+  InitializePhase initializePhase() const override { return InitializePhase::Primary; }
+  static absl::StatusOr<std::unique_ptr<DnsClusterImpl>>
+  create(const envoy::config::cluster::v3::Cluster& cluster,
+         const envoy::extensions::clusters::dns::v3::DnsCluster& dns_cluster,
+         ClusterFactoryContext& context, Network::DnsResolverSharedPtr dns_resolver);
+
+protected:
```

Comment:
Is there a reason why constructor has to be protected instead of just private?

I figure that protected only matters when something derives from this class, which should not be the case given that it's new and nothing in this PR inherits it.